client.go

```diff
@@ -785,13 +785,22 @@ func (f *File) Name() string {
 // than calling Read multiple times. io.Copy will do this
 // automatically.
 func (f *File) Read(b []byte) (int, error) {
+	r, err := f.ReadAt(b, int64( f.offset ))
+	f.offset += uint64(r)
+	return r, err
+}
+
+// ReadAt reads up to len(b) byte from the File at a given offset `off`. It returns 
+// the number of bytes read and an error, if any. ReadAt follows io.ReaderAt semantics, 
+// so the file offset is not altered during the read.
+func (f *File) ReadAt(b []byte, off int64) (n int, err error) {
```

Comment:
From the documentation about the `io.ReaderAt` semantics:
```
When ReadAt returns n < len(p), it returns a non-nil error explaining why more bytes were not returned. In this respect, ReadAt is stricter than Read.

If some data is available but not len(p) bytes, ReadAt blocks until either all the data is available or an error occurs. In this respect ReadAt is different from Read.
```

It looks like unfortunately, the proper semantics between `io.ReaderAt` means that it cannot be immediately used with an underlying tracking offset as an `io.Reader` like you are doing. 🙁 It’s just close enough to be different.

Reply:
I agree that the io.ReaderAt has different requirements than io.Reader, reading the io.Reader seems to indicate that io.ReaderAt semantics are still valid io.Reader semantics.

So io.ReaderAt specifies...

> If some data is available but not len(p) bytes, ReadAt blocks until either all the data is available or an error occurs. In this respect ReadAt is different from Read.

While io.Reader says...

> If some data is available but not len(p) bytes, Read conventionally returns what is available instead of waiting for more.

I read 'conventionally' here to mean that it can, but doesn't have to, return instead of waiting. That is it is still valid Reader semantics to wait for more data instead of returning thus the ReaderAt behavior is still valid Reader behavior.

The other differing semantics of ReaderAt...

>When ReadAt returns n < len(p), it returns a non-nil error explaining why more bytes were not returned. In this respect, ReadAt is stricter than Read.

If it does wait for more data, then it seems to me that it would fulfill this as well as it would either eventually return the full data or partial data and an error.

If correct then io.ReaderAt semantics is valid io.Reader semantics. So this would all be OK except that we don't want to change the existing behavior. But after a quick reading of the client Read() method, it looks to me like it actually has the more restrictive, io.ReaderAt like semantics. It doesn't stop sending SFTP requests and reading in the returned data until it has filled up the passed in slice (b) or hit an error.

@puellanivis, @qrpike if either/both of you would read through the Read() methods and confirm my reading I'd appreciate it. If it is correct, then this might work.

Reply:
So, including having read your assessment here, my understanding is that `ReaderAt` is stricter than `Reader` so it _ought_ to work for all uses of `Reader`. So, all we should need to do is ensure the new `ReadAt` method fits the requirements of `ReaderAt` and then it should be good.

Slightly varying specifications are the most enjoyable part of the work, no? :wink:

Reply:
> my understanding is that ReaderAt is stricter than Reader so it ought to work for all uses of Reader

Not only should it work but Reader, as currently implemented, already has the semantics of io.ReaderAt except for the missing offset parameter. I do need to read through it one more time to double check that the semantics do line up.